src/lib/ext-monitor.js

```diff
@@ -0,0 +1,110 @@
+import { getActivityLogPageURL } from './ext-listen.js';
+
+export default class ExtensionMonitor {
+  logs = [];
+  // Map<string, Function>
+  extensionMapList = new Map([]);
+
+  async getAllExtensions() {
+    const extensions = await browser.management.getAll();
+    const self = await browser.management.getSelf();
+    return extensions.filter((extension) => {
+      return extension.type === 'extension' && extension.id !== self.id;
+    });
+  }
+
+  async isActivityLogPageOpen() {
+    const tab = await browser.tabs.query({
+      url: getActivityLogPageURL(),
+    });
+    return tab.length > 0;
+  }
+
+  // If the Activity Log page is open, each logs will be send to
+  // Activity Log page (as soon as it is encountered).
+  async sendLogs(details) {
+    const isExtPageOpen = await this.isActivityLogPageOpen();
+    if (isExtPageOpen) {
+      await browser.runtime.sendMessage({
+        requestType: 'appendLogs',
+        log: details,
+      });
+    }
+  }
+
+  createLogListener() {
+    return async (details) => {
+      this.logs.push(details);
+      await this.sendLogs(details);
+    };
+  }
+
+  hasActivityListeners() {
+    return this.extensionMapList.size > 0;
+  }
+
+  async startMonitor() {
+    if (!this.hasActivityListeners()) {
+      const extensions = await this.getAllExtensions();
+      return new Promise((resolve) => {
+        extensions.forEach((extension) => {
+          const listener = this.createLogListener();
+          this.extensionMapList.set(extension.id, listener);
+          browser.activityLog.onExtensionActivity.addListener(
+            listener,
+            extension.id
+          );
+        });
+        resolve(true);
+      });
+    } else {
+      throw new Error('EAM is already running');
+    }
+  }
+
+  async stopMonitor() {
+    return new Promise((resolve) => {
+      this.extensionMapList.forEach((listener, extensionId) => {
+        browser.activityLog.onExtensionActivity.removeListener(
+          listener,
+          extensionId
+        );
+        this.extensionMapList.delete(extensionId);
+      });
+      resolve(true);
+    });
+  }
+
+  messageHandlers = {
+    getMonitorStatus: () => {
+      return {
+        active: this.hasActivityListeners(),
+      };
+    },
+    startMonitor: () => {
+      return this.startMonitor();
+    },
+    stopMonitor: () => {
+      return this.stopMonitor();
+    },
+    sendAllLogs: () => {
+      return {
+        existingLogs: this.logs,
+      };
+    },
+  };
+
+  messageListener = async (message) => {
+    const { requestType } = message;
+
+    if (requestType in this.messageHandlers) {
+      return this.messageHandlers[requestType]();
+    } else {
+      throw new Error('requestType ' + requestType + ' is not found');
+    }
+  };
```

Comment:
I've been thinking about this a bit more after I suggested you that this could have been an async function yesterday:

at the moment this may be fine because this is the only `browser.runtime.onMessage` listener that the extension is registering, but at some point this may become confusing because the WebExtensions internals will assume that this messageListener is always handling any message it receives (because it always return a promise).

We may consider this a follow up to take care of in one of the next pull request, but let's keep it in mind.

To avoid this we may do something like this:
- do not make this function as async
- include another property that make it explicitly which module is supposed to handle the message (e.g. a `requestTo` property)
- return earlier if the message wasn't meant to be handled by this message listener

which would look like more or less like this:
```
messageListener = async (message) => {
  const { requestType, requestTo } = message;
  if (requestTo !== "ext-monitor) {
    return;
  }

  if (requestType in this.messageHandlers) {
    try {
      return Promise.resolve(this.messageHandlers[requestType]());
    } catch (err) {
      return Promise.reject(err);
    }
  } else {
    return Promise.reject(new Error('...'));
  }
}
```

and then include `requestTo: 'ext-monitor'` in the methods defined in ext-listen.js that are sending the control message that this module should handle.